rcl/include/rcl/time.h

```diff
@@ -28,52 +28,384 @@ extern "C"
 #define RCL_MS_TO_NS(milliseconds) (milliseconds * 1000 * 1000)
 #define RCL_US_TO_NS(microseconds) (microseconds * 1000)
 
-#define RCL_NS_TO_S(nanoseconds) (nanoseconds / 1000 * 1000 * 1000)
-#define RCL_NS_TO_MS(nanoseconds) (nanoseconds / 1000 * 1000)
+#define RCL_NS_TO_S(nanoseconds) (nanoseconds / (1000 * 1000 * 1000))
+#define RCL_NS_TO_MS(nanoseconds) (nanoseconds / (1000 * 1000))
 #define RCL_NS_TO_US(nanoseconds) (nanoseconds / 1000)
 
-/// Struct which encapsulates a point in time according to a system clock.
-/* The system clock's point of reference is the Unix Epoch (January 1st, 1970).
- * See: http://stackoverflow.com/a/32189845/671658
- * Therefore all times in this struct are positive and count the nanoseconds
- * since the Unix epoch and this struct cannot be used on systems which report
- * a current time before the Unix Epoch.
- * Leap seconds are not considered.
- *
- * The struct represents time as nanoseconds in an unsigned 64-bit integer.
- * The struct is capable of representing any time until the year 2554 with
- * nanosecond precisions.
- */
-typedef struct rcl_system_time_point_t
+
+enum rcl_time_source_type_t
 {
-  uint64_t nanoseconds;
-} rcl_system_time_point_t;
-
-/// Struct which encapsulates a point in time according to a steady clock.
-/* The steady clock's point of reference is system defined, but often the time
- * that the process started plus a signed random offset.
- * See: http://stackoverflow.com/a/32189845/671658
- * However, the clock is guaranteed to be monotonically increasing.
- * Therefore all times in this struct are positive and count nanoseconds
- * since an unspecified, but constant point in time.
- *
- * The struct represents time as nanoseconds in an unsigned 64-bit integer.
- */
-typedef struct rcl_steady_time_point_t
+  RCL_TIME_SOURCE_UNINITIALIZED = 0,
+  RCL_ROS_TIME,
+  RCL_SYSTEM_TIME,
+  RCL_STEADY_TIME
+};
+
+typedef uint64_t rcl_time_point_imp_t;
+typedef int64_t rcl_duration_imp_t;
+
+typedef struct rcl_time_source_t
 {
-  uint64_t nanoseconds;
-} rcl_steady_time_point_t;
+  enum rcl_time_source_type_t type;
+  void (* pre_update)(void);
+  void (* post_update)(void);
+  rcl_ret_t (* get_now)(void * data, rcl_time_point_imp_t * now);
+  // void (*set_now) (rcl_time_point_imp_t);
+  void * data;
+} rcl_time_source_t;
+
+struct rcl_ros_time_source_storage_t;
+typedef struct rcl_time_point_t
+{
+  rcl_time_point_imp_t ns;
+  rcl_time_source_t * time_source;
+} rcl_time_point_t;
 
-/// Struct which encapsulates a duration of time between two points in time.
-/* The struct can represent any time within the range [~292 years, ~-292 years]
- * with nanosecond precision.
- */
 typedef struct rcl_duration_t
 {
-  int64_t nanoseconds;
+  rcl_time_point_imp_t ns;
+  rcl_time_source_t * time_source;
 } rcl_duration_t;
 
-/// Retrieve the current time as a rcl_system_time_point_t struct.
+// typedef struct rcl_rate_t
+// {
+//   rcl_time_point_imp_t trigger_time;
+//   int64_t period;
```

Comment:
Yes, I commented this out as out, to remove it from the scope for this PR. There's a todo at the bottom to come back to this struct